### PR TITLE
fix: prevent short-circuit during env variable substitution

### PR DESCRIPTION
**Pull request by alexmt**
PR fixes issue reported in https://github.com/argoproj/argo-cd/pull/3202 and prevents incorrect env variable replacement.

Signed-off-by: Alexander Matyushentsev <AMatyushentsev@gmail.com>

### CODE REVIEW

**Review by jannfis — APPROVED** [review]
LGTM, albeit it seems to have a little fragile touch to me due to the regexp quirks.

Also, I was wondering for the reasoning about leaving unset variables "as is" instead of expanding them to the empty string.

**Review by jgwest — APPROVED** [review]
LGTM, I wasn't able to poke any logic/test holes in the code (excepting those noted by @jannfis on Slack/PR). @jannfis questioned whether we should expand unset var to the empty string, I'd be curious what user's expectations are here (I could make the argument either way, but would assume that expanding the unset var to empty string is more likely to be expected)

**Comment by alexmt** [comment]
Great question about why not expand missing env variables to the empty string. It was discussed before https://github.com/argoproj/argo-cd/pull/2415#discussion_r331979270. Looks like there is no strong reason not to do it. Updating PR

File: docs/operator-manual/upgrading/1.8-1.9.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# v1.8 to 1.9
```
Comment (alexmt) [inline docs/operator-manual/upgrading/1.8-1.9.md:1]: Intentionally not mentioning upgrade instructions in TOC to avoid confusion.

Comment (jannfis) [reply docs/operator-manual/upgrading/1.8-1.9.md:1]: I like the idea to use this document for tracking changes that need attention during upgrade together with the PR that introduces such a change. We should keep this habit :+1:

**Review by jannfis — APPROVED** [review]
Great & proper, easy to maintain fix @alexmt!

LGTM

**Pull request merged**